### PR TITLE
[ISSUE-897] Add ANONYMOUS Permission 

### PR DESCRIPTION
**Pull request by seonwoo960000**
### Modification 
- Add `ProjectRole#ANONYMOUS` 
- Modify UI 
image
image

### Result 
- Users will be able to add `ANONYMOUS` role to their project 

### Links 
https://github.com/line/centraldogma/issues/897

### CODE REVIEW

**Comment by minwoox** [comment]
>  I wonder how we can allow anonymous users to access public repositories(without passing access token) with the Authorizer's token checking logic.

Let's create an `anonymous` user and set the user if the token is `anoymouse` in the authorizer.

File: server/src/test/java/com/linecorp/centraldogma/server/metadata/MetadataServiceTest.java [hunk 1]
```diff
@@ -182,6 +182,8 @@ void perRolePermissions() {
                                                                                      Permission.WRITE);
         assertThat(repositoryMetadata.perRolePermissions().guest()).containsExactly(Permission.READ,
                                                                                     Permission.WRITE);
+        assertThat(repositoryMetadata.perRolePermissions().anonymous()).containsExactly(Permission.READ,
+                                                                                        Permission.WRITE);
```
[the remark this answers is not in the document]
Comment (seonwoo960000) [reply server/src/test/java/com/linecorp/centraldogma/server/metadata/MetadataServiceTest.java:186]: Sure, I'll remove the permission 👍

**Review by thachlp — APPROVED** [review]

**Comment by seonwoo960000** [comment]
> Let's create an anonymous user and set the user if the token is anoymouse in the authorizer.

@minwoox, please check if I've understood correctly 😄 
- Add an anonymous user in `com.linecorp.centraldogma.server.metadata.User`
- Users will be able to add tokens with App Id of "anonymous". The `Authorizer` will set `Anonymous` user when token's App Id is "anonymous"

**Comment by minwoox** [comment]
> Users will be able to add tokens with App Id of "anonymous".

Does "Users" mean the clients or the repository owners?

The repository owners do not have to add the tokens. All they have to do is adding the read permission to the anonymous role. Here is the flow:
- Add `ANONONYMOUS` to `User`:
  ```java
  public class User {
    public static final User ANONYMOUS = ...
  ```
- Clients set the `Authrorizaion: anonymous` header to the request.
- The authorizer sets the anonymous user to the request context using `AuthUtil.setCurrentUser`.
  - https://github.com/line/centraldogma/blob/9ecbb3201c3a399c133d537b180da63019d8e33c/server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/ApplicationTokenAuthorizer.java#L59
- `RequiresPermissionDecorator` checks the permission:
  - https://github.com/line/centraldogma/blob/2b07420e92f99011d7ee459ee96ac5cc1d035038/server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/RequiresPermissionDecorator.java#L48
  - https://github.com/line/centraldogma/blob/3ddf67bc7c45189c0fc4cd634dade527ad19579b/server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java#L766-L777
  
 Please let me know if you have any further questions. 😉

**Comment by seonwoo960000** [comment]
@minwoox, thank you so much for the detailed explanation. It helped me a lot 👍 

It seems like `com.linecorp.centraldogma.client.armeria.AbstractCentralDogmaBuilder` sets `com.linecorp.centraldogma.internal.CsrfToken#ANONYMOUS` as accessToken by default. So the question is, should we use `CsrfToken#ANONYMOUS` when checking anonymous token, or should we create a new `ApplicationToken#ANONYMOUS` for the purpose of checking anonymous token?

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/ApplicationTokenAuthorizer.java [hunk 2]
```diff
@@ -57,6 +59,11 @@ public ApplicationTokenAuthorizer(Function<String, CompletionStage<Token>> token
     @Override
     public CompletionStage<Boolean> authorize(ServiceRequestContext ctx, HttpRequest data) {
         final OAuth2Token token = AuthTokenExtractors.oAuth2().apply(data.headers());
+        if (token != null && token.accessToken().equals(ApplicationToken.ANONYMOUS)) {
```
Comment (seonwoo960000) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/ApplicationTokenAuthorizer.java:62]: @minwoox, thank you so much for the detailed explanation. It helped me a lot 👍 

It seems like `com.linecorp.centraldogma.client.armeria.AbstractCentralDogmaBuilder` sets `com.linecorp.centraldogma.internal.CsrfToken#ANONYMOUS` as accessToken by default. So the question is, should we use `CsrfToken#ANONYMOUS` when checking anonymous token, or should we create a new `ApplicationToken#ANONYMOUS` for the purpose of checking anonymous token?

File: common/src/main/java/com/linecorp/centraldogma/internal/ApplicationToken.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.linecorp.centraldogma.internal;
+
+/**
+ * Holds a default application token.
+ */
+public class ApplicationToken {
```
Comment (seonwoo960000) [inline common/src/main/java/com/linecorp/centraldogma/internal/ApplicationToken.java:6]: @minwoox, thank you so much for the detailed explanation. It helped me a lot 👍

It seems like com.linecorp.centraldogma.client.armeria.AbstractCentralDogmaBuilder sets com.linecorp.centraldogma.internal.CsrfToken#ANONYMOUS as accessToken by default. So the question is, should we use CsrfToken#ANONYMOUS when checking anonymous token, or should we create a new ApplicationToken#ANONYMOUS for the purpose of checking anonymous token?

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/ApplicationTokenAuthorizer.java (hunk 2 above)
Comment (seonwoo960000) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/ApplicationTokenAuthorizer.java:62]: Use `ApplicationToken.ANONYMOUS` or `CsrfToken.ANONYMOUS`?

File: common/src/main/java/com/linecorp/centraldogma/internal/ApplicationToken.java (hunk 3 above)
Comment (seonwoo960000) [reply common/src/main/java/com/linecorp/centraldogma/internal/ApplicationToken.java]: I've added a commit that uses `CsrfToken#ANONYMOUS` and remove `ApplicationToken`

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java [hunk 4]
```diff
@@ -765,9 +766,10 @@ private CompletableFuture<Collection<Permission>> findPermissions0(String projec
             final RepositoryMetadata repositoryMetadata = metadata.repo(repoName);
             final Member member = metadata.memberOrDefault(user.id(), null);
 
-            // If the member is guest.
+            // If the member is guest or using anonymous token.
             if (member == null) {
-                return repositoryMetadata.perRolePermissions().guest();
+                return !user.isAnonymous() ? repositoryMetadata.perRolePermissions().guest() :
+                       repositoryMetadata.perRolePermissions().anonymous();
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:772]: nit:
```suggestion
                return !user.isAnonymous() ? repositoryMetadata.perRolePermissions().guest()
                                           : repositoryMetadata.perRolePermissions().anonymous();
```

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java [hunk 5]
```diff
@@ -73,14 +73,14 @@ public static PerRolePermissions ofDefault() {
      * Creates a {@link PerRolePermissions} which allows accessing a repository from everyone.
      */
     public static PerRolePermissions ofPublic() {
-        return new PerRolePermissions(READ_WRITE, READ_WRITE, READ_WRITE);
+        return new PerRolePermissions(READ_WRITE, READ_WRITE, READ_WRITE, READ_ONLY);
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java:76]: Let's remove the read permission to anonymouse users by default. 😉 
```suggestion
        return new PerRolePermissions(READ_WRITE, READ_WRITE, READ_WRITE, NO_PERMISSION);
```

File: server/src/test/java/com/linecorp/centraldogma/server/internal/api/auth/PermissionTest.java [hunk 6]
```diff
@@ -184,6 +189,22 @@ void test(@Nullable String appId, String secret, String projectName, ProjectRole
         }
     }
 
+    @Test
+    void test_anonymous() {
+        final WebClient client = WebClient.builder(server.httpUri())
```
Comment (minwoox) [inline server/src/test/java/com/linecorp/centraldogma/server/internal/api/auth/PermissionTest.java:194]: nit: we can use the blocking client.
```
final BlockingWebClient client = server.blockingWebClient(
                builder -> builder.addHeader(HttpHeaderNames.AUTHORIZATION, "Bearer " + CsrfToken.ANONYMOUS));
```

**Review by minwoox — APPROVED** [review]
Looks great! Thanks! 👍 👍 👍

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java [hunk 7]
```diff
@@ -413,9 +413,10 @@ public CompletableFuture<Revision> updatePerRolePermissions(Author author,
...
         if (Project.REPO_META.equals(repoName)) {
             final Set<Permission> guest = perRolePermissions.guest();
-            if (!guest.isEmpty()) {
+            final Set<Permission> anonymous = perRolePermissions.anonymous();
+            if (!guest.isEmpty() || !anonymous.isEmpty()) {
                 throw new UnsupportedOperationException(
-                        "Can't give a permission to guest for internal repository: " + repoName);
+                        "Can't give a permission to guest or anonymous for internal repository: " + repoName);
             }
         }
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:422]: Could you also check if the anonymous permission has the write permission? If so, we should raise an exception.

File: server/src/test/java/com/linecorp/centraldogma/server/internal/api/auth/PermissionTest.java (hunk 6 above)
Comment (seonwoo960000) [reply server/src/test/java/com/linecorp/centraldogma/server/internal/api/auth/PermissionTest.java]: Oh, there was a better way! Thanks 👍

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java (hunk 7 above)
Comment (seonwoo960000) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:425]: I've fixed to throw exception when trying to grant WRITE permission to anonymous user

**Review by jrhee17 — APPROVED** [review]
Changes look reasonable to me 👍   Thanks @seonwoo960000 !

nit; we may need to update https://github.com/line/centraldogma/blob/2994712c79f041941d5fb5160368823b59832e29/site/src/sphinx/auth.rst?plain=1#L293 later if not now

**Review by trustin — APPROVED** [review]
Excellent work, @seonwoo960000! 🙇

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java [hunk 8]
```diff
@@ -98,16 +98,23 @@ public static PerRolePermissions ofPrivate() {
...
     @JsonCreator
     public PerRolePermissions(@JsonProperty("owner") Iterable<Permission> owner,
                               @JsonProperty("member") Iterable<Permission> member,
-                              @JsonProperty("guest") Iterable<Permission> guest) {
+                              @JsonProperty("guest") Iterable<Permission> guest,
+                              @JsonProperty("anonymous") Iterable<Permission> anonymous) {
         this.owner = Sets.immutableEnumSet(requireNonNull(owner, "owner"));
         this.member = Sets.immutableEnumSet(requireNonNull(member, "member"));
         this.guest = Sets.immutableEnumSet(requireNonNull(guest, "guest"));
+        this.anonymous = Sets.immutableEnumSet(requireNonNull(anonymous, "anonymous"));
```
Comment (ikhoon) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java:117]: Is this compatible with the existing data? `anonymous` may be null if `PerRolePermissions` is serialized like:
```json
"perRolePermissions": {
  "owner": [
    "READ",
    "WRITE"
  ],
  "member": [
    "READ",
    "WRITE"
  ],
  "guest": [
    "READ",
    "WRITE"
  ]
}
```

Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java]: Thanks! Addressed. 😉

**Review by ikhoon — APPROVED** [review]
Thanks, @seonwoo960000! 🙇‍♂️🙇‍♂️ This would be a go-to feature for many users.

**Comment by minwoox** [comment]
@seonwoo960000 👍 👍 👍

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java (hunk 8 above)
Comment (seonwoo960000) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java]: Oh, how was it addressed 🤔 ?

Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/PerRolePermissions.java]: Here it is 😉 : https://github.com/line/centraldogma/pull/917/commits/bb55dcabc6701dddb633c8c6a48b81eee7860f9b

**Pull request merged**